### PR TITLE
feat(settings): Make suggestion providers much more configurable at runtime

### LINKED ISSUE #120: Allow full configuration of Merino providers
Currently, Merino's providers are configured using limited presets. This means that much of the flexibility in the system is only configurable by making code changes. We should allow using the full flexibility from the YAML config files. 

The config does not need to be accessible from environment variables.

**Acceptance Criteria**

* Any number of providers can be configured in the YAML file
* Providers are arbitrarily nestable, for providers that support nesting (like caches)
* Top level providers should be named, to enable future toggling by clients

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/CONSVC-1244)

### PR DESCRIPTION
**Pull request by mythmon**
- Remove Cows and lifetimes from providers and box all of them
- feat(settings): Make suggestion providers much more configurable at runtime

fixes #120 

BREAKING CHANGE: The configuration of providers is more free-form now, and the old settings will need updated.

### CODE REVIEW

**Comment by mythmon** [comment]
@hackebrot I made some changes to the `test-engineering` tests, since the configuration system changed enough that it wasn't compatible anymore.

**Review by hackebrot — COMMENTED** [review]
Hi @mythmon! 👋🏻 I added my comments inline.

**Comment by mythmon** [comment]
> Looks like the system tests failed on CI with a timeout connecting to Merino:

By default `wait-for-it.sh` will still run the command if it times out. This is similar to something I had hit locally. I can work on this, though if you'd prefer to make changes yourself, please go ahead. I'll be at RustConf today, so won't get a chance to look at this right way.

**Comment by hackebrot** [comment]
I was looking into the timeout issues again today. The log message below suggests that Merino is running on a port `8080` even though `config/ci.yaml` sets it to `8000`:

config:
```yml
http:
  # The host and port Merino listens on
  listen: "0.0.0.0:8000"
```

log:
```
INFOactix_server::builder: Starting "actix-web-service-0.0.0.0:8080" service on 0.0.0.0:8080
```

**Comment by mythmon** [comment]
Well that's not good. I'll look into that and see what's going on.

File: test-engineering/README.md [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 **Please note that this is work in progress.** 🚧
 
 ## Run tests locally
 
-You can run the integration tests locally from the repository root directory using:
+You can run the system tests locally from the repository root directory using:
 
 ```text
-docker compose -f test-engineering/docker-compose.yml up --abort-on-container-exit --build
+docker-compose -f test-engineering/docker-compose.yml up --abort-on-container-exit --build
```
[the remark this answers is not in the document]
Comment (mythmon) [reply test-engineering/README.md:12]: Ah, interesting. My version of docker does not have a compose subcommand, and this is a typo I make very often, so I assumed it was the same.

File: Dockerfile [hunk 2]
```diff
@@ -79,10 +79,6 @@ COPY --from=builder /app/bin /app/bin
 COPY --from=builder /app/version.json /app
 COPY --from=builder /app/config /app/config
 
-ARG HOST=0.0.0.0
-ARG PORT=8080
-ENV MERINO_HTTP__LISTEN="${HOST}:${PORT}"
```
Comment (mythmon) [inline Dockerfile:84]: This was the real fix. The rest of the changes in this commit are useful things I figured out while debugging this.

**Comment by hackebrot** [comment]
The integration tests job failed with the following error:

```text
merino_client | pytest: error: unrecognized arguments: merino:8000 -- pytest -vv
```

Looks like something is wrong with the command for the `merino_client` Docker service and how `/wait-for-it.sh` forwards arguments. 🤔

**Comment by mythmon** [comment]
@hackebrot btw, every commit of the form `area: description` goes into the changelog. For incremental fixes to a branch, I don't think we should use them. I like JR's pattern of denoting these fixup-commits that don't contribute to the changelog (and should be squashed or considered squashed) by writing `f description` as in "fixup".

**Comment by mythmon** [comment]
Rebased to fix merge conflicts

**Comment by mythmon** [comment]
@mozilla-services/context-services This should be ready for review now that @hackebrot fixed up the tests.

**Review by pjenvey — APPROVED** [review]

**Pull request merged**